Implement CloudnativePG database operator

pvannierop opened the pull request:
**Description of the change**
This PR will deploy the CloudNativePG postgresql/timescaledb operator for Kubernetes. CloudNativePG will deploy a new database cluster that can be managed (migrated, scaled, ...) by the CNPG. All RADAR-base services that make use of a database are reconfigured. Most of this is done at the level op the helm charts. Where relevant logic/config of database passwords was changed or removed. Migration documentation is included in the PR.

**Benefits**
The CloudNativePG postgresql/timescaledb operator will make managing the database a lot more efficient.

**Possible drawbacks**
To make the transition, users may have to migrate their databases. CloudNativePG operator automates this to some degrees (see migration documentation).

keyvaann commented on helmfile.d/05-cloudnativepg-operator.yaml:
```diff
@@ -0,0 +1,18 @@
+  - name: cloudnative-pg
+    namespace: administration
+#    chart: radar/radar-cloudnative-pg
+    chart: ../../radar-helm-charts/external/cloudnative-pg/cloudnative-pg
```
This should be updated

keyvaann commented on helmfile.d/05-cloudnativepg-operator.yaml:
```diff
@@ -0,0 +1,18 @@
+    chart: ../../radar-helm-charts/external/cloudnative-pg/cloudnative-pg
+    version: {{ .Values.cloudnative_pg._chart_version }}
+    installed: {{ .Values.cloudnative_pg._install }}
+    timeout: {{ add .Values.base_timeout .Values.kube_prometheus_stack._extra_timeout }}
```
This shouldn't default to `kube_prometheus_stack`

keyvaann commented on helmfile.d/10-services.yaml:
```diff
@@ -288,62 +302,64 @@ releases:
         value: {{ .Values.server_name }}
 
   - name: kratos
```
It's good that name of the chart and name of the release to match, otherwise during deployment you'll have a pod named `radar-kratos-kratos` which unnecessarily long. If both chart name and release name match the pod will named something like `radar-kratos`

keyvaann commented on etc/base.yaml:
```diff
@@ -178,17 +178,40 @@ postgresql:
+cloudnative_postgresql:
+  _install: true
+  _chart_version: 0.1.0
+  _extra_timeout: 0
```
It's good to allow configuration of the storage here since it is something that will be changed more often.

keyvaann replied:
Allowing configuration of cluster size would also be good here

pvannierop replied:
I added these config options as was already present for the timescaledb instances

keyvaann commented on etc/base.yaml:
```diff
@@ -178,17 +178,40 @@ postgresql:
+  _chart_version: 0.1.0
+  _extra_timeout: 0
+  needs:
+    - cloudnative-pg
```
I'm not sure if this should be here.

pvannierop replied:
I have tried deploying the cnpg-operator as a helm dependency. This fails because of order problems with the CRD installation. I have now reworked this by 1) changing the creation condition of the operator and 2) keep the `needs:` entry, all the helmfile level, 3) remove the separate _05-..._ helmfile. It cannot get better than this I am afraid.

keyvaann commented on etc/base.yaml:
```diff
@@ -178,17 +178,40 @@ postgresql:
         - name: POSTGRES_MULTIPLE_DATABASES
           value: restsourceauthorizer,appconfig,kratos # ... and these databases will be created
 
```
Also maybe add a note to the postgresql configuration above, mentioning that it is deprecated.

pvannierop replied:
I added deprecation warnings as well as corrected comments around database deployments

keyvaann commented on etc/base.yaml:
```diff
@@ -178,17 +178,40 @@ postgresql:
+  _extra_timeout: 0
+  needs:
+    - cloudnative-pg
+#  Uncomment this when migrating from the legacy postgresql databases
```
What would happen if these steps are enabled by default?

pvannierop replied:
It would try to ingest data from a non existing upstream source. Most likely, erroring out the `initdb` pod that is started. Would you argue to have this always on?

keyvaann commented on etc/base.yaml:
```diff
@@ -294,11 +314,6 @@ radar_rest_sources_backend:
   _chart_version: 1.2.3
   _extra_timeout: 0
   replicaCount: 1
-  postgres:
```
I think it's good to instead of deleting these values to move them to a file in the mod directory and and documentation on how to use the RADAR-Kubernetes stack while using a managed PostgreSQL server such as RDS.

pvannierop replied:
Sounds good, but perhaps nice to add as a Issue, rather than an addition to this PR?

pvannierop committed:
feat: remove database secret generation

pvannierop committed:
feat: remove postgresql passwords from base-secrets.yaml

pvannierop committed:
feat: implement cloudnative-pg charts

pvannierop committed:
fix: username and password for grafana metrics endpoint

pvannierop committed:
refactor: create clean production.yaml

Production.yaml no longer repeats base chart versions (and other items) defined in the RADAR-Kubenetes release. From here on it will only be used to indicate deviations from base.yaml.

pvannierop committed:
docs: add cloudnative-pg migration and external database configuration instructions

pvannierop committed:
ci: update e2e-test configuration

yatharthranjan reviewed (approved):
Thanks for this. I had a quick look and looks good to me.

When releasing, can you please make sure to put `BREAKING CHANGES` at the top of the release description and a link to the migration docs.

mpgxvii commented:
@pvannierop Thanks for this. Would we be able to use the existing configurations if we are using managed services (like AWS RDS)?

pvannierop commented on etc/base.yaml:
```diff
@@ -373,50 +372,76 @@ realtime_dashboard_timescaledb:
+    enabled: true
+    cluster:
+      cluster:
+        instances: 2
```
@yatharthranjan I have set the default to 1 replicate
